### PR TITLE
Wilson slack variables

### PR DESCRIPTION
**Pull request by wilson-ricks**
For all the major policy constraints (Capacity Reserve Margin, CO2 Cap, Energy Share Requirement, and Minimum Capacity Requirement), the presence of the appropriate input files will now cause GenX to create slack variables for the relevant constraint. A slack variable will allow the specified constraint to be violated while incurring a specified objective function penalty. This will prevent GenX from throwing infeasibilities when policy constraints are violated, and can allow users to set price caps beyond which policy constraints no longer apply. Slack variables will only be created if the required input files are present. Examples of the required files and formats are present in the new example system SmallNewEngland/ThreeZones_Slack_Variables_Example. Output files will be written for any slack variables present.

### CODE REVIEW

**Review by sambuddhac — APPROVED** [review]
Reviewed and approved !!!

File: src/load_inputs/load_minimum_capacity_requirement.jl [hunk 1]
```diff
@@ -28,5 +28,11 @@ function load_minimum_capacity_requirement!(path::AbstractString, inputs::Dict,
     if setup["ParameterScale"] == 1
         inputs["MinCapReq"] /= ModelScalingFactor # Convert to GW
     end
+    if "PriceCap" in names(df)
+        inputs["MinCapPriceCap"] = df[!,:PriceCap]
+        if setup["ParameterScale"] == 1
```
Comment (cfe316) [inline src/load_inputs/load_minimum_capacity_requirement.jl:33]: This function now has two `if setup["ParameterScale"]` blocks. We could replace both with
`scale_factor = setup["ParameterScale"] == 1 ? ModelScalingFactor : 1`.

Comment (wilson-ricks) [reply src/load_inputs/load_minimum_capacity_requirement.jl:33]: Yeah, that would be a useful change.

File: src/write_outputs/co2_cap/write_co2_cap.jl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+@doc raw"""
+	write_co2_cap(path::AbstractString, inputs::Dict, setup::Dict, EP::Model)
+
+Function for reporting carbon price associated with carbon cap constraints.
+
+"""
+function write_co2_cap(path::AbstractString, inputs::Dict, setup::Dict, EP::Model)
+    dfCO2Price = DataFrame(CO2_Cap = [Symbol("CO2_Cap_$cap") for cap = 1:inputs["NCO2Cap"]],
+                            CO2_Price = (-1) * (dual.(EP[:cCO2Emissions_systemwide])))
+    if setup["ParameterScale"] == 1
```
Comment (cfe316) [inline src/write_outputs/co2_cap/write_co2_cap.jl:26]: This function has two `if setup["ParameterScale"] == 1` blocks. They could be replaced using

`    scale_factor = setup["ParameterScale"] == 1 ? ModelScalingFactor : 1`

File: src/write_outputs/energy_share_requirement/write_esr_prices.jl [hunk 3]
```diff
@@ -19,6 +19,15 @@ function write_esr_prices(path::AbstractString, inputs::Dict, setup::Dict, EP::M
 	if setup["ParameterScale"] == 1
 		dfESR[!,:ESR_Price] = dfESR[!,:ESR_Price] * ModelScalingFactor # Converting MillionUS$/GWh to US$/MWh
 	end
-	CSV.write(joinpath(path, "ESR_prices.csv"), dfESR)
+
+	if haskey(inputs, "dfESR_slack")
+		dfESR[!,:ESR_AnnualSlack] = convert(Array{Float64}, value.(EP[:vESR_slack]))
+		dfESR[!,:ESR_AnnualPenalty] = convert(Array{Float64}, value.(EP[:eCESRSlack]))
+		if setup["ParameterScale"] == 1
```
Comment (cfe316) [inline src/write_outputs/energy_share_requirement/write_esr_prices.jl:26]: This function has two `if setup["ParameterScale"] == 1` blocks. They could be replaced using

`scale_factor = setup["ParameterScale"] == 1 ? ModelScalingFactor : 1`

**Comment by cfe316** [comment]
Is this PR also actually two PRs? One for slack variables and one for multi-stage? Can they be separated?

**Comment by wilson-ricks** [comment]
This was built off of a branch where I'd implemented the multi-stage changes, which I believe has already been accepted in a separate PR

**Pull request merged**